File: java/142_Linked_List_Cycle_II/142_Linked_List_Cycle_II.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        HashSet<ListNode> set = new HashSet<ListNode>();
        ListNode cur = head;
        int pos = 0;
        while(cur != null) {
            if (!set.contains(cur.next)) {
                set.add(cur);
                cur = cur.next;
                pos++;
            } else {
                break;
            }
        }
        return cur != null ? cur.next : cur;
    }
}

File: go/142_Linked_List_Cycle_II/142_Linked_List_Cycle_II.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
    m := make(map[*ListNode]int)
    cur := head
    pos := 0
    for cur != nil {
        _, exists := m[cur]
        if !exists {
            m[cur] = cur.Val
            cur = cur.Next
            pos++
        } else {
            break
        }
    }
    return cur
}

